qiskit/transpiler/synthesis/graysynth.py

```diff
@@ -0,0 +1,304 @@
+
+
+def graysynth(cnots, angles, n_sections=2):
+    """
+    This function is an implementation of the GraySynth algorithm.
+
+    GraySynth is a heuristic algorithm for synthesizing small parity networks.
+    It is inspired by Gray codes. Given a set of binary strings S
+    (called "cnots" bellow), the algorithm synthesizes a parity network for S by
+    repeatedly choosing an index i to expand and then effectively recursing on
+    the co-factors S_0 and S_1, consisting of the strings y in S,
+    with y_i = 0 or 1 respectively. As a subset S is recursively expanded,
+    CNOT gates are applied so that a designated target bit contains the
+    (partial) parity ksi_y(x) where y_i = 1 if and only if y'_i = 1 for for all
+    y' in S. If S is a singleton {y'}, then y = y', hence the target bit contains
+    the value ksi_y'(x) as desired.
+
+    Notably, rather than uncomputing this sequence of CNOT gates when a subset S
+    is finished being synthesized, the algorithm maintains the invariant
+    that the remaining parities to be computed are expressed over the current state
+    of bits. This allows the algorithm to avoid the 'backtracking' inherent in
+    uncomputing-based methods.
+
+    The algorithm is described in detail in the following paper in section 4:
+    "On the controlled-NOT complexity of controlled-NOT–phase circuits."
+    Amy, Matthew, Parsiad Azimzadeh, and Michele Mosca.
+    Quantum Science and Technology 4.1 (2018): 015002.
+
+    Args:
+        cnots (list[list]): a matrix whose columns are the parities to be synthesized
+            e.g.
+                [[0, 1, 1, 1, 1, 1],
+                 [1, 0, 0, 1, 1, 1],
+                 [1, 0, 0, 1, 0, 0],
+                 [0, 0, 1, 0, 1, 0]]
+            corresponds to:
+                 x1^x2 + x0 + x0^x3 + x0^x1^x2 + x0^x1^x3 + x0^x1
+
+        angles (list): a list containing all the phase-shift gates which are
+            to be applied, in the same order as in "cnots". A number is
+            interpreted as the angle of u1(angle), otherwise the elements
+            have to be 't', 'tdg', 's', 'sdg' or 'z'.
+
+        n_sections (int): number of sections, used in _lwr_cnot_synth(), in the
+            Patel–Markov–Hayes algorithm. n_sections must be a factor of n_qubits.
+
+    Returns:
+        QuantumCircuit: the quantum circuit
+
+    Raises:
+        QiskitError: when dimensions of cnots and angles don't align
+    """
+    n_qubits = len(cnots)
+
+    # Create a quantum circuit on n_qubits
+    qcir = QuantumCircuit(n_qubits)
+
+    if len(cnots[0]) != len(angles):
+        raise QiskitError('Size of "cnots" and "angles" do not match.')
+
+    range_list = list(range(n_qubits))
+    epsilon = n_qubits
+    sta = []
+    cnots_copy = np.transpose(np.array(copy.deepcopy(cnots)))
+    state = np.eye(n_qubits).astype('int')  # This matrix keeps track of the state in the algorithm
+
+    # Check if some T-gates can be applied, before adding any C-NOT gates
```

Comment:
I don't quite follow this part of the code, what is the benefit of adding the phase gates early?

Reply:
The idea of this part of the code is just to apply those phase-shift gates before the actual algorithms starts (it should say phase-shift gates, not T-gates...I fixed this now). It is just there to speed things up, as the whole algorithm then doesn't have to deal with such trivial gates